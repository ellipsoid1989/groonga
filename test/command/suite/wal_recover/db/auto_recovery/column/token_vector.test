# _database_close/_database_reopen can't use with HTTP because requests via HTTP are always
# processed by workers. Workers never use primary WAL role.
#@require-interface stdio

#@require-feature message_pack

#$GROONGA_ALLOW_DATABASE_REOPEN=yes
#$GRN_WAL_ROLE=primary

table_create Terms TABLE_PAT_KEY ShortText \
  --normalizer NormalizerNFKC130 \
  --default_tokenizer TokenNgram

table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText
column_create Data content_terms COLUMN_VECTOR Terms content

load --table Data
[
{"_key": "abc",    "content": "a and b and c"},
{"_key": "ABC",    "content": "A and B and C"},
{"_key": "abcdef", "content": "a and b and c and d and e and f"}
]

delete --table Data --key ABC

dump

_database_close

#@copy-path #{db_path}.0000103 #{db_path}.0000103.wal
#@add-important-log-levels notice
_database_reopen
#@remove-important-log-levels notice

select Data

dump
