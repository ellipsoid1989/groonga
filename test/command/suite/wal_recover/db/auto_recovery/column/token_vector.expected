table_create Terms TABLE_PAT_KEY ShortText   --normalizer NormalizerNFKC130   --default_tokenizer TokenNgram
[[0,0.0,0.0],true]
table_create Data TABLE_PAT_KEY ShortText
[[0,0.0,0.0],true]
column_create Data content COLUMN_SCALAR ShortText
[[0,0.0,0.0],true]
column_create Data content_terms COLUMN_VECTOR Terms content
[[0,0.0,0.0],true]
load --table Data
[
{"_key": "abc",    "content": "a and b and c"},
{"_key": "ABC",    "content": "A and B and C"},
{"_key": "abcdef", "content": "a and b and c and d and e and f"}
]
[[0,0.0,0.0],3]
delete --table Data --key ABC
[[0,0.0,0.0],true]
dump
table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenNgram --normalizer NormalizerNFKC130

column_create Data content_terms COLUMN_VECTOR Terms content

load --table Data
[
["_key","content"],
["abc","a and b and c"],
["abcdef","a and b and c and d and e and f"]
]

load --table Terms
[
["_key"],
["a"],
["and"],
["b"],
["c"],
["d"],
["e"],
["f"]
]
_database_close
[[0,0.0,0.0],true]
_database_reopen
[[0,0.0,0.0],true]
#|e| [reader][read-entry][ja][recover][Data.content_terms] must be map: <positive-integer>(2): path:<db/db.0000103>
#|n| [db][wal][recover] rebuild broken column: <Data.content_terms>(259)
#|n| DDL:260:column_create Data #recovering#content_terms
#|n| spec:260:update:Data.#recovering#content_terms:65(column:var_size):256(Terms)
#|n| DDL:260:set_source Data.#recovering#content_terms Data.content
#|n| spec:258:update:Data.content:65(column:var_size):14(ShortText)
#|n| spec:260:update:Data.#recovering#content_terms:65(column:var_size):256(Terms)
#|n| DDL:259:obj_remove Data.content_terms
#|n| spec:258:update:Data.content:65(column:var_size):14(ShortText)
#|n| spec:259:remove:Data.content_terms:65(column:var_size)
#|n| [db][wal][recover] succeeded to rebuild broken column: <Data.content_terms>(260)
select Data
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        2
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "content",
          "ShortText"
        ],
        [
          "content_terms",
          "Terms"
        ]
      ],
      [
        1,
        "abc",
        "a and b and c",
        [
          "a",
          "and",
          "b",
          "and",
          "c"
        ]
      ],
      [
        3,
        "abcdef",
        "a and b and c and d and e and f",
        [
          "a",
          "and",
          "b",
          "and",
          "c",
          "and",
          "d",
          "and",
          "e",
          "and",
          "f"
        ]
      ]
    ]
  ]
]
dump
table_create Data TABLE_PAT_KEY ShortText
column_create Data content COLUMN_SCALAR ShortText

table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenNgram --normalizer NormalizerNFKC130

column_create Data content_terms COLUMN_VECTOR Terms content

load --table Data
[
["_key","content"],
["abc","a and b and c"],
["abcdef","a and b and c and d and e and f"]
]

load --table Terms
[
["_key"],
["a"],
["and"],
["b"],
["c"],
["d"],
["e"],
["f"]
]
